bindings/sundials4py/cvodes/cvodes.cpp

```diff
@@ -0,0 +1,345 @@
+                      nb::arg("cvode_mem"), nb::arg("efun").none());
+
+  BIND_CVODE_CALLBACK(CVodeSetNlsRhsFn, CVRhsFn, fNLS, cvode_nlsrhsfn_wrapper,
+                      nb::arg("cvode_mem"), nb::arg("f").none());
+
+  BIND_CVODE_CALLBACK(CVodeSetJacFn, CVLsJacFn, lsjacfn, cvode_lsjacfn_wrapper,
+                      nb::arg("cvode_mem"), nb::arg("jac").none());
+
+  BIND_CVODE_CALLBACK2(CVodeSetPreconditioner, CVLsPrecSetupFn, lsprecsetupfn,
+                       cvode_lsprecsetupfn_wrapper, CVLsPrecSolveFn,
+                       lsprecsolvefn, cvode_lsprecsolvefn_wrapper,
+                       nb::arg("cvode_mem"), nb::arg("pset").none(),
+                       nb::arg("psolve").none());
+
+  BIND_CVODE_CALLBACK2(CVodeSetJacTimes, CVLsJacTimesSetupFn, lsjactimessetupfn,
+                       cvode_lsjactimessetupfn_wrapper, CVLsJacTimesVecFn,
+                       lsjactimesvecfn, cvode_lsjactimesvecfn_wrapper,
+                       nb::arg("cvode_mem"), nb::arg("jtsetup").none(),
+                       nb::arg("jtimes").none());
+
+  BIND_CVODE_CALLBACK(CVodeSetLinSysFn, CVLsLinSysFn, lslinsysfn,
+                      cvode_lslinsysfn_wrapper, nb::arg("cvode_mem"),
+                      nb::arg("linsys").none());
+
+  BIND_CVODE_CALLBACK(CVodeSetJacTimesRhsFn, CVLsJacTimesVecFn, lsjacrhsfn,
+                      cvode_lsjacrhsfn_wrapper, nb::arg("cvode_mem"),
+                      nb::arg("jtimesRhsFn").none());
+
+  BIND_CVODE_CALLBACK(CVodeSetProjFn, CVProjFn, projfn, cvode_projfn_wrapper,
+                      nb::arg("cvode_mem"), nb::arg("pfun").none());
+
+  m.def("CVodeQuadSensInit",
+        [](void* cv_mem, std::function<CVQuadSensRhsStdFn> fQS,
+           std::vector<N_Vector> yQS0)
+        {
+          auto fn_table = get_cvode_fn_table(cv_mem);
+          fn_table->fQS = nb::cast(fQS);
+          return CVodeQuadSensInit(cv_mem, cvode_fQS_wrapper, yQS0.data());
+        });
+
+  m.def("CVodeSensInit",
+        [](void* cv_mem, int Ns, int ism, std::function<CVSensRhsStdFn> fS,
+           std::vector<N_Vector> yS0)
+        {
+          auto fn_table = get_cvode_fn_table(cv_mem);
+          fn_table->fS  = nb::cast(fS);
+          return CVodeSensInit(cv_mem, Ns, ism, cvode_fS_wrapper, yS0.data());
+        });
+
+  m.def("CVodeSensInit1",
+        [](void* cv_mem, int Ns, int ism,
+           std::function<std::remove_pointer_t<CVSensRhs1Fn>> fS1,
+           std::vector<N_Vector> yS0)
+        {
+          auto fn_table = get_cvode_fn_table(cv_mem);
+          fn_table->fS1 = nb::cast(fS1);
+          return CVodeSensInit1(cv_mem, Ns, ism, cvode_fS1_wrapper, yS0.data());
+        });
+
+  ///
+  // CVODES Adjoint Bindings
+  ///
+
+  m.def("CVodeInitB",
+        [](void* cv_mem, int which,
+           std::function<std::remove_pointer_t<CVRhsFnB>> fB, sunrealtype tB0,
+           N_Vector yB0)
+        {
+          int cv_status = CVodeInitB(cv_mem, which, cvode_fB_wrapper, tB0, yB0);
```

Comment:
Since we always pass `cvode_fB_wrapper` check for `NULL` input and check return
```suggestion
          if (!fB) { throw sundials4py::illegal_value("rhs was null"); }

          int cv_status = CVodeInitB(cv_mem, which, cvode_fB_wrapper, tB0, yB0);
          if (cv_status != CV_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize CVODE");
          }
```

Reply:
Check `cv_status` and throw if needed

Reply:
Should we throw or return the error code?